### PR TITLE
Feature toggles

### PR DESCRIPTION
**Pull request by JohannesDoberer**

### CODE REVIEW

File: docs/advanced-scenarios.md [hunk 1]
```diff
@@ -152,4 +152,61 @@ Google's `id_token` contains basic identity data like name and user ID, which al
...
+}
+```
+
+#### Use feature toggles in a micro frontend.
+It is also possible to restrict content in a micro frontend using feature toggles. The active feature toggle list is available in the Luigi [client api](luigi-client-api.md#getActiveFeatureToggles).
+```javascript
+  if (LuigiClient.getActiveFeatureToggles().includes('ft1')) {
+    //display content
+  }
+```
```
Comment (alexandra-simeonova) [inline docs/advanced-scenarios.md:211]:
`````suggestion

#### Usage

* To **set** feature toggles, you have two possibilities:

  1. Set feature toggles to the active feature toggle list through [Luigi Core API](luigi-core-api.md#featuretoggles):
  ```javascript
  Luigi.featureToggles().setFeatureToggle('ft1');
  ```

  2. Set feature toggles to the active feature toggle list via URL parameters:
  ```
  http://localhost:4200/projects/pr1?ft=ft1,ft2
  ```

* To **unset** feature toggles, you have to use the Core API:

  ```javascript
    Luigi.featureToggles().unsetFeatureToggle('ft1');
  ```

* To **restrict node visiblity with feature toggles**:

  You can define a list of feature toggles for a particular top or left navigation node. Then, you can use the [visibleForFeatureToggles](navigation-parameters-reference.md#visibleForFeatureToggles) parameter in order to display the node for certain feature toggles.
  For example, this node will be visible if `ft1` is added to the active feature toggle list:

  ```javascript
  {
      category: { label: 'Feature Toggle: Settings 2', icon: 'action-settings' },
      pathSegment: 'settings_ft',
      label: 'Project Settings 2',
      viewUrl: '/sampleapp.html#/projects/' + projectId + '/settings',
      icon: 'settings',
      visibleForFeatureToggles: ['ft1']
  }
  ```

  It is also possible to negate the visibility of a node by adding an exclamation mark at the beginning of the feature toggle name.
  In this example, the node will not be visible if `ft1` is set as an active feature toggle, but if another feature toggle like `ft2` is active, the node will be displayed:
  ```javascript
  {
      category: { label: 'Feature Toggle: Settings 2', icon: 'action-settings' },
      pathSegment: 'settings_ft',
      label: 'Project Settings 2',
      viewUrl: '/sampleapp.html#/projects/' + projectId + '/settings',
      icon: 'settings',
      visibleForFeatureToggles: ['!ft1']
  }
  ```

* To **use feature toggles in a micro frontend**:

  It is possible to restrict content in a micro frontend using feature toggles. The active feature toggle list is available in the Luigi [Client API](luigi-client-api.md#getActiveFeatureToggles).
  ```javascript
    if (LuigiClient.getActiveFeatureToggles().includes('ft1')) {
      //display content
    }
  ```
`````

File: docs/advanced-scenarios.md [hunk 2]
```diff
@@ -152,4 +152,61 @@ Google's `id_token` contains basic identity data like name and user ID, which al
...
+#### Set feature toggles to the active feature toggle list via url
+```
+  http://localhost:4200/projects/pr1?ft=ft1,ft2
+```
+
+In order to remove a feature toggle from the list you have to use the core api
+
+#### unset feature toggles from the active feature toggle list through core api
+```javascript
+  Luigi.featureToggles().setFeatureToggle('ft1');
```
Comment (alexandra-simeonova) [inline docs/advanced-scenarios.md:173]: I assume it should be **unsetFeatureToggle** here instead, so I changed it in my suggestion (let me know if it's ok)

File: docs/navigation-parameters-reference.md [hunk 3]
```diff
@@ -305,6 +305,21 @@ settings: {
   - **errorFn** used to handle different scenarios other than redirection.
   - **since**: v1.0.1
 
+### visibleForFeatureToggles
+- **type**: array
+- **description**: defines a list of feature toggles valid for this node. This list will be compared against the active feature toggles list. If there is a match the node will be visible in the left navigation. It is also possible to negate the feature Toggle. That means if a feature toggle, defined in this list, starts with exclamation mark it will not be visible if the same feature toggle is set in the active feature toggle list but for all other feature toggles.
```
Comment (alexandra-simeonova) [inline docs/navigation-parameters-reference.md:310]:
```suggestion
- **description**: defines a list of feature toggles valid for this node. This list will be compared against the active feature toggles list. If there is a match, the node will be visible in the left navigation. It is also possible to negate the feature toggle by adding a `!` in front of the toggle name.  Then, the node will be visible for all other feature toggles except the one defined with an exclamation point.
```

File: docs/advanced-scenarios.md [hunk 4]
```diff
@@ -152,4 +152,61 @@ Google's `id_token` contains basic identity data like name and user ID, which al
...
+```
+
+Displaying nodes restricted to feature toggles in the top or left navigation you can define a list of feature toggles on the node.
```
Comment (ndricimrr) [inline docs/advanced-scenarios.md:176]: - This sentence seems like it belongs beneath the title below: _#### Visibility of a node restricted by feature toggles_ .  Over here seems like it belongs to _“unset feature toggles..."_ right above. 

- Somethings a bit off with the sentence. How about sth like: **_In order to add node restrictions through feature toggles to the top or left navigation nodes you can define a list of feature toggles on the node_** , probably @alexandra-simeonova  can come up with sth better

File: docs/advanced-scenarios.md [hunk 5]
```diff
@@ -152,4 +152,61 @@ Google's `id_token` contains basic identity data like name and user ID, which al
...
+
+#### Visibility of a node restricted by feature toggles
+If there is a feature toggle `ft1` added to the active feature toggle list, the node will be displayed when [visibleForFeatureToggles](navigation-parameters-reference.md#visibleForFeatureToggles) is defined.
```
Comment (ndricimrr) [inline docs/advanced-scenarios.md:179]:
```suggestion
If there is a feature toggle `ft1` added to the active feature toggle list, the node will be displayed when [visibleForFeatureToggles](navigation-parameters-reference.md#visibleForFeatureToggles) is defined and also contains `ft1` as well.
```

File: test/e2e-test-application/src/app/project/project.component.html [hunk 6]
```diff
@@ -291,6 +291,17 @@ <h3 class="fd-layout-panel__title">Navigate</h3>
               data="linkManager().navigate('users/groups/stakeholders')"
             ></app-code-snippet>
           </li>
+          <li class="fd-list__item">
+            <a
+              href="javascript:void(0)"
+              class="fd-link"
+              (click)="linkManager().navigate('/projects/pr1/settings?ft=test')"
```
Comment (ndricimrr) [inline test/e2e-test-application/src/app/project/project.component.html:298]: This worked very well, just that after navigation finished you don't really notice what happens unless you query `getActiveFeatureToggleList`. Was thinking to add one more line to print/show a text if `ft=test` on the settings page as well, just like you do for ft1 🤔

File: test/e2e-test-application/src/app/project/project.component.html [hunk 7]
```diff
@@ -359,6 +370,24 @@ <h3 class="fd-layout-panel__title">Navigate</h3>
...
           </li>
+          <li class="fd-list__item">
+            <a
+              href="javascript:void(0)"
+              class="fd-link"
+              (click)="
+                linkManager()
+                  .fromClosestContext()
+                  .withParams({ feature: 'toggle' })
+                  .navigate('settings_ft?ft=test')
```
Comment (ndricimrr) [inline test/e2e-test-application/src/app/project/project.component.html:381]: I got an error when clicking on this node:

Screenshot 2020-08-20 at 08 21 43

File: core/src/services/routing.js [hunk 8]
```diff
@@ -168,9 +168,21 @@ class RoutingClass {
...
+      const featureToggleProperty = LuigiConfig.getConfigValue(
+        'settings.featureToggles.queryStringParam'
+      )
+        ? LuigiConfig.getConfigValue('settings.featureToggles.queryStringParam')
+        : undefined;
+      if (featureToggleProperty) {
+        RoutingHelpers.setFeatureToggles(featureToggleProperty, path);
+      } else {
+        console.warn(
+          'featureToggles.queryStringParam is not defined in settings.js'
```
[the remark this answers is not in the document]
Comment (JohannesDoberer) [reply core/src/services/routing.js:180]: true.. good point

File: core/src/App.html [hunk 9]
```diff
@@ -459,9 +463,12 @@
...
       path = Routing.concatenatePath(getSubPath(currentNode), params.link);
     }
-
     if (params.nodeParams && Object.keys(params.nodeParams).length) {
-      path += '?';
+      if (path.includes('?')) {
+        path += '&';
+      } else {
+        path += '?';
+      }
```
[the remark this answers is not in the document]
Comment (maxmarkus) [reply core/src/App.html]: works, imho harder to read as the original solution.

File: core/src/core-api/featuretoggles.js [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+   */
+  unsetFeatureToggle(featureToggleName) {
+    if (featureToggleName && typeof featureToggleName === 'string') {
+      if (get(this.featureToggleList).includes(featureToggleName)) {
+        let index = get(this.featureToggleList).indexOf(featureToggleName);
+        if (index > -1) {
+          get(this.featureToggleList).splice(index, 1);
+        }
+      } else {
+        console.log('Feature toggle name is not in the list.');
```
Comment (maxmarkus) [inline core/src/core-api/featuretoggles.js:42]:
```suggestion
        console.warn('Feature toggle name is not in the list.');
```

File: core/src/core-api/featuretoggles.js [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+      }
+    } else {
+      console.log('Feature toggle name is empty or not a type of string');
```
Comment (maxmarkus) [inline core/src/core-api/featuretoggles.js:45]:
```suggestion
      console.warn('Feature toggle name is empty or not a type of string');
```
I think we should avoid using classical logs, but use more meaningful console functions.

File: core/src/core-api/featuretoggles.js [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+   * @memberof FeatureToggles
+   * @since NEXTRELEASE
+   * @example Luigi.featureToggles().setFeatureToggle('featureToggleName');
+   */
+  setFeatureToggle(featureToggleName) {
+    if (featureToggleName && typeof featureToggleName === 'string') {
+      if (!get(this.featureToggleList).includes(featureToggleName)) {
+        get(this.featureToggleList).push(featureToggleName);
+      } else {
+        console.log('Feature toggle name already exists');
```
Comment (maxmarkus) [inline core/src/core-api/featuretoggles.js:21]:
```suggestion
        console.warn('Feature toggle name already exists');
```

File: core/src/core-api/featuretoggles.js [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+      }
+    } else {
+      console.log('Feature toggle name is empty or not a type of string');
```
Comment (maxmarkus) [inline core/src/core-api/featuretoggles.js:24]:
```suggestion
      console.warn('Feature toggle name is empty or not a type of string');
```

File: test/e2e-test-application/src/app/project/settings/settings.component.html [hunk 14]
```diff
@@ -93,3 +93,11 @@ <h3 class="fd-layout-panel__title">
     </div>
   </div>
 </section>
+<section
+  *ngIf="testFeatureToggleActive"
+  class="fd-section luigi_featuretoggle_ft1"
```
Comment (maxmarkus) [inline test/e2e-test-application/src/app/project/settings/settings.component.html:98]: Is the class _luigi_featuretoggle_ft1_ used somehow? Couldn't find a reference.

File: website/docs/src/luigi-config/extended/routing.js [hunk 15]
```diff
@@ -6,7 +6,7 @@ class Routing {
    * For hash routing, set to true and run `npm run start`
    * For path routing, set to false and run `npm run start`
    */
-  useHashRouting = false;
+  useHashRouting = true;
```
Comment (maxmarkus) [inline website/docs/src/luigi-config/extended/routing.js:9]: The docs should run with path routing?
```suggestion
  useHashRouting = false;
```

File: core/src/core-api/featuretoggles.js [hunk 16]
```diff
@@ -0,0 +1,60 @@
...
+
+  /**
+   * Get a list of active feature toggles
+   * @memberof FeatureToggles
+   * @since NEXTRELEASE
+   * @return {Array} of active feature toggles
+   * @example Luigi.featureToggles().getActiveFeatureToggleList();
+   */
+  getActiveFeatureToggleList() {
+    return get(this.featureToggleList);
```
Comment (maxmarkus) [inline core/src/core-api/featuretoggles.js:57]: Would be awesome to avoid possible mutations by returning a clone instead of reference.

**Review by maxmarkus — CHANGES_REQUESTED** [review]
Good job, just some remarks.

**Review by maxmarkus — APPROVED** [review]
LGTM 👍

**Pull request merged**